domains/client/domain-executor/src/tests.rs

```diff
@@ -1234,3 +1235,198 @@ async fn test_cross_domains_message_should_work() {
+        alice
+            .get_open_channel_for_domain(DomainId::CORE_PAYMENTS)
+            .is_some()
+            && bob.get_open_channel_for_domain(DomainId::SYSTEM).is_some()
+    })
+    .await
+    .unwrap();
+
+    // Register `charlie` as relayer such that message will assign to it, but as its relayer
+    // is not started these massage won't be relayed.
+    bob.construct_and_send_extrinsic(pallet_messenger::Call::join_relayer_set {
+        relayer_id: Charlie.into(),
+    })
+    .await
+    .expect("Failed to construct and send extrinsic");
+    produce_blocks!(ferdie, alice, bob, charlie, 3)
+        .await
+        .unwrap();
+
+    // Create cross domain message, only message assigned to `alice` and `bob` will be relayed
+    // and send to tx pool, and these message is unordered because the message assigned to `charlie`
+    // is not relayed.
+    let relayer_id: AccountId = Charlie.into();
+    let alice_transfer_amount = 1;
+    let bob_transfer_amount = 2;
+    let pre_alice_free_balance = alice.free_balance(alice.key.to_account_id());
+    let pre_bob_free_balance = bob.free_balance(bob.key.to_account_id());
+    let mut alice_account_nonce = alice.account_nonce();
+    let mut bob_account_nonce = bob.account_nonce();
+    // Assigne `inbox_response` message to `charlie`
+    for _ in 0..10 {
+        let tx = alice.construct_extrinsic(
+            alice_account_nonce,
+            pallet_transporter::Call::transfer {
+                dst_location: pallet_transporter::Location {
+                    domain_id: DomainId::CORE_PAYMENTS,
+                    account_id: AccountIdConverter::convert(Bob.into()),
+                },
+                amount: alice_transfer_amount,
+            },
+        );
+        alice
+            .send_extrinsic(tx)
+            .await
+            .expect("Failed to send extrinsic");
+        alice_account_nonce += 1;
+
+        produce_blocks!(ferdie, alice, bob, charlie, 1)
+            .await
+            .unwrap();
+    }
+    // Assigne `outbox` message to `charlie`
+    for _ in 0..10 {
+        let tx = bob.construct_extrinsic(
+            bob_account_nonce,
+            pallet_transporter::Call::transfer {
+                dst_location: pallet_transporter::Location {
+                    domain_id: DomainId::SYSTEM,
+                    account_id: AccountIdConverter::convert(Alice.into()),
+                },
+                amount: bob_transfer_amount,
+            },
+        );
+        bob.send_extrinsic(tx)
+            .await
+            .expect("Failed to send extrinsic");
+        bob_account_nonce += 1;
+
+        produce_blocks!(ferdie, alice, bob, charlie, 1)
+            .await
+            .unwrap();
+    }
+
+    // Run charlie's relayer worker, the message assigned to `charlie` will be relayed
+    // and send to tx pool now
+    let relayer_worker = domain_client_message_relayer::worker::relay_core_domain_messages::<
```

Comment:
IIUC, when a tx arrives if its `requires` is fulfilled (either by another tx in the ready queue or it has no `requires`) it will be added to the TX pool's ready queue, and other tx in the future queue that fulfilled by this tx's `provides` will also be added to the ready queue.

If a tx's `requires` is not fulfilled when it arrived it will be added into the TX pool's future queue and wait for another tx that fulfilled its `provides` to be added to the ready queue.